Added ability to configure react-server-cli with a .reactserverrc file or through package.json

aickin opened the pull request:
@gigabo mentioned in [a comment](https://github.com/redfin/react-server/pull/49#issuecomment-205476978) on PR #49 that it would be nice if there was a config file way to launch `react-server-cli`. I agree, and that's what this PR does.

If you send in any arguments at the command line, nothing changes. But if you don't, then it starts looking in the current directory for either a `.reactserverrc` file or a `reactServer` section of package.json. If it doesn't find either of those, it starts crawling up the directory hierarchy to find a config file. It uses the first config file it finds; it doesn't try to find multiple configs and merge them. For any values that aren't specified, it uses the same defaults it used before.

Additionally, as in Babel, it allows for config overrides on a per-environment basis. All you do is include a `env` attribute in your config, like this:

```json
{
  "port": 4000,
  "env": {
    "production": 5000
  }
}
```

In this example, the port will be 5000 when NODE_ENV is set to "production" and 4000 otherwise.

Other notes:

* I got rid of the `--production` flag; it seemed like it was excessively confusing to me, both at the command line and in terms of the code. It also felt like setting NODE_ENV to production is a fairly normal thing these days, and we should just go with that convention and not provide a second way to do it. Let me know if you disagree.
* I moved the logging warnings from the CLI-specific code into `startServer` so that they will happen for API uses as well as CLI uses.
* One side effect of this PR is that the API now has different config defaults when NODE_ENV is production, which was not true before. This is by design, but I wanted to call it out.

gigabo commented:
> a `.reactserverrc` 

Aw yeah!

> a `reactServer` section of package.json

Huh, cool.  I have mixed feelings about tucking config into `package.json` generally.  I see the appeal of having package config in one place.  On the other hand small single-purpose configuration files provide good separation of concerns.  Not opposed to this, and it seems to be a trend.  Just voicing my skepticism. :)

> If you send in any arguments at the command line, nothing changes.

This seems unusual.  More typical behavior with configuration files is to allow command-line options to override values _individually_, right?  Whereas here if I add _any_ command-line option (e.g. `--log-level=debug`) I'm then responsible for also re-specifying everything that would otherwise have been picked up from my config?  I'd prefer a merge sieve of `defaults => config file => options`.

> it doesn't try to find multiple configs and merge them.

This is another common pattern.  Not sure if it makes sense for `react-server` or not, but it's a logical extension of per-option command-line config overrides.

> config overrides on a per-environment basis.

Awesome!  I don't get the example, though.  I would have expected the value associated with `production` to be an object with key/value configuration pairs rather than a single value:

```javascript
{
  "port": 4000,
  "env": {
    "production": {
      "port": 5000
    }
  }
}
```

aickin commented:
> I have mixed feelings about tucking config into package.json generally... it seems to be a trend. Just voicing my skepticism.

I hear that. I was basing it off `babel` (which you probably knew), and I thought it was kind of neat, but I'm not wedded to it. It's very removable if you want.

> More typical behavior with configuration files is to allow command-line options to override values individually, right?

I'm actually not sure which is more common; I'll investigate with babel.

What I was worried about is reproducibility of a particular CLI invocation. It just felt potentially confusing to me to have the same command line set of options do different things depending on your working directory. (Of course, I recognize at some level that having `react-server-cli` work differently in different directories is **the entire point** of this PR, but it somehow felt more constrained when it only applied to option-less invocations of the CLI. I'm not sure that's rational on my behalf.) 

I also was thinking that a single config's values probably have some interdependencies (which is why I also didn't want to merge config from different sources in the file system), but I may be wrong about that, too. Your example of just changing the log level is illuminating.

I'll look at babel's behavior and think about it a little more. Implementation should be easy no matter what the decision.

> Awesome! I don't get the example, though.

Oh, yeah, the example is just plain wrong. D'oh. Your version is correct, as is the example in the README.md. Thanks for catching it.

aickin commented:
Played around with babel, and it follows your proposal for merging (`defaults => config file => options`). I've also just thought about it more, and I think you are right and my original idea was wrong. 

Implemented the new options merging in 4120e21, which actually meant I got to delete a bit of code. Yay!

As for `package.json` config, in the absence of major resistance, I'm going to leave it in. If you change your mind and want it out, though, just let me know.

Thanks as always for the feedback!

gigabo commented on packages/react-server-cli/README.md:
```diff
@@ -43,11 +43,39 @@ It's rare to see a project these days in the JavaScript world that isn't at leas
-1. Otherwise, the base mode is **development**.
+###Ways to add options
+
+There are three ways to pass options to the CLI, through the command line, `.reactserverrc` JSON files, or as a `reactServer` entry in `package.json` files. Once `react-server` finds one of these sources of options, it will not search for more. It searches for options this way:
```
I think this wording isn't quite right anymore with your latest changes.  Command-line options no longer cause `react-server` to stop searching for additional sources of options.  Same with the pseudo-code below.

gigabo commented on packages/react-server-cli/src/defaultOptions.js:
```diff
@@ -0,0 +1,16 @@
+// these are the default options for startServer.
```
Nice to have this pulled out into a dedicated module!  Also nice to have consistent defaults between the CLI and a direct call `startServer`.

gigabo commented on packages/react-server-cli/src/findOptionsInFiles.js:
```diff
@@ -0,0 +1,41 @@
+				return parsedPackageJson.reactServer;
+			}
+		}
+	} while (dir !== (dir = path.dirname(dir)))
```
> do { ... } while (dir !== (dir = path.dirname(dir)))

Yeah! 😈

gigabo commented on packages/react-server-cli/src/parseCliArgs.js:
```diff
@@ -9,52 +9,56 @@ export default (isProduction, args = process.argv) => {
+		}
+	}
+
+	return result;
```
What's up with the explicit `undefined` => remove `undefined` dance?

aickin replied:
Values that aren't sent in on the command line come in as `undefined` (except for boolean values, which come in as `false`.

Sending in a value with `undefined` overrides the default, because I used `Object.assign` to merge options, so I remove them before sending in to startServer.

I'll comment it.

gigabo commented on packages/react-server-cli/src/startServer.js:
```diff
@@ -3,19 +3,34 @@ import http from "http"
-		hot = true,
-		minify = false,
-		compileOnly = false,
+export default (routesRelativePath, options = {}) => {
```
Ahhh... That's nice. 😌

gigabo commented on packages/react-server-cli/src/startServer.js:
```diff
@@ -128,3 +143,33 @@ const handleCompilationErrors = (err, stats) => {
+		logging.setLevel('time',  'fast');
+		logging.setLevel('gauge', 'ok');
+	}
+}
```
I like that this happens when I call `startServer` directly now.  I'd like it _more_ if it happened in the _browser_, too! ;)

gigabo commented:
This is awesome!

Next question: Can this merge with [this config](https://github.com/redfin/react-server/blob/master/packages/react-server/core/config.js)?  Should it?

Some values here probably shouldn't be exposed to the client.  Feels weird to have _two_ config files, though...

aickin commented:
Changed the README and added comments on the CLI `undefined` dance. Thanks!

> Next question: Can this merge with this config? Should it?

Oh hey. I'd completely forgotten about that file's existence (even though blame says I wrote significant parts of it... sigh). I can't easily figure out what was in that config; do you know, or do I need to go code spelunking?

gigabo commented:
I don't think _anything_ is put in it or used from it by `react-server` itself.  But `react-server` transfers it automagically from the server to the browser, which is convenient.  We use it for things like asset versions and server domains.  I guess all stuff that can be transferred via a separate script include?

@roblg - Any reason not to just kill it?

aickin commented:
> I don't think anything is put in it or used from it by react-server itself.
> @roblg - Any reason not to just kill it?

If that's the case, how would you feel about filing it as a bug and separating it from this PR? It seems like it's not something that most folks writing app code need to know about?

roblg commented:
> Any reason not to just kill it?

Only the reasons that you already outlined -- mainly that it's available both on the client and server. I think it's baked into the lifecycle in `ClientController` a little bit, but I can't imagine it'd be that hard to replace w/ a separate script.

gigabo commented:
> filing it as a bug and separating it from this PR

[Done](https://github.com/redfin/react-server/issues/91).
